In [1]:
import findspark
findspark.init('/usr/local/spark')
from pyspark.sql import SparkSession

In [2]:
import pycountry
from resources.update_local_data import get_covid_data

In [3]:
spark = SparkSession.builder.appName("covid_deaths").getOrCreate()

Update local data

In [4]:
get_covid_data()

[INFO] url read -> https://covid.ourworldindata.org/data/owid-covid-data.csv
[INFO] data/covid_deaths.csv saved
[INFO] data/covid_vaccinations.csv saved


# Covid-19 Dataset from Our World in Data
- main dataset source: https://ourworldindata.org/covid-deaths
- complement dataset, population density per country: https://worldpopulationreview.com/country-rankings/countries-by-density

## Overview:
The present notebooks tends to explore and analyze covid deaths dataset from Our World in Data. It going to take special attention to three countries, Soain, Italy and USA, while looking for some global numbers and doing a general analysis.

It can be noted that PySpark is going to be used for this analysis, actually it's going to be taken advantage of SQL syntax thank to `spark.sql()` for complex queries.

## Steps:
1. Load dataset, and check its schema, structure and scope,
2. Look at the most infected countries, where are they located (which continent),
3. Look at the countries with more deaths, where are they located (which continent),
4. Compare the previous results with Spain, Italy and USA,

### 1. Load dataset, and check its schema, structure and scope
- covid_deaths.csv
- world_population_density.csv

In [5]:
df_deaths = spark.read.csv('../data/covid_deaths.csv', inferSchema=True, header=True)
df_pop_density = spark.read.csv('../data/world_population_density.csv', inferSchema=True, header=True)

#### covid_deaths.csv

In [6]:
df_deaths.createOrReplaceTempView("covid_deaths")
df_deaths.printSchema()

root
 |-- _c0: integer (nullable = true)
 |-- iso_code: string (nullable = true)
 |-- continent: string (nullable = true)
 |-- location: string (nullable = true)
 |-- date: string (nullable = true)
 |-- total_cases: double (nullable = true)
 |-- new_cases: double (nullable = true)
 |-- new_cases_smoothed: double (nullable = true)
 |-- total_deaths: double (nullable = true)
 |-- new_deaths: double (nullable = true)
 |-- new_deaths_smoothed: double (nullable = true)
 |-- total_cases_per_million: double (nullable = true)
 |-- new_cases_per_million: double (nullable = true)
 |-- new_cases_smoothed_per_million: double (nullable = true)
 |-- total_deaths_per_million: double (nullable = true)
 |-- new_deaths_per_million: double (nullable = true)
 |-- new_deaths_smoothed_per_million: double (nullable = true)
 |-- reproduction_rate: double (nullable = true)
 |-- icu_patients: double (nullable = true)
 |-- icu_patients_per_million: double (nullable = true)
 |-- hosp_patients: double (nullable = 

Continents that appear in the dataset

In [7]:
spark.sql("""
    SELECT DISTINCT continent
    FROM covid_deaths
""").show()

+-------------+
|    continent|
+-------------+
|       Europe|
|       Africa|
|         null|
|North America|
|South America|
|      Oceania|
|         Asia|
+-------------+



There are 231 countries from 6 different continents

In [8]:
spark.sql("""
    SELECT DISTINCT location
    FROM covid_deaths
""").count()

230

This is the range of the dataset. The records start on january 2020 and should end today since it's udtated everytime the function `update_local_data` run at the beginning of this notebook

In [9]:
spark.sql("""
    SELECT min(date(date)) as Start, max(date(date)) as End
    FROM covid_deaths
""").show()

+----------+----------+
|     Start|       End|
+----------+----------+
|2020-01-01|2021-06-29|
+----------+----------+



#### world_population_density.csv

In [10]:
df_pop_density.createOrReplaceTempView("world_population_density")
df_pop_density.printSchema()

root
 |-- rank: integer (nullable = true)
 |-- country: string (nullable = true)
 |-- density: double (nullable = true)
 |-- densityMi: double (nullable = true)
 |-- pop2021: double (nullable = true)
 |-- area: integer (nullable = true)



world_population_density.csv has 232 countries contemplated, one more than covid_deaths.csv

In [11]:
df_pop_density.select('country').count()

232

Now, let's compare these two tables or datasets:
- `covid_deaths.csv`
- `world_population_density.csv`

The idea is to enhance covid_deaths dataset by adding the population density per country. So let's compare which countries in `covid_deaths.csv` appear in `world_population_density.csv`.

For better result, it can be taken advantage of the iso_code standard, so that the language or any local convention can't affect when two countries name are compared (e.g. Czechia and Czech Republic refer to the same country).

In the case of `covid_deaths` table, it already have the iso_code as the second column, but the `world_population_density` just have the country name. Therefore, for this second case, it's going to be used `pycountry` (a pyhotn package available here: https://pypi.org/project/pycountry/) to decode the country name to effectively compare the two datasets.

More specifically, it's going to be used map transformation of `PySpark` that recieve a function as parameters to later applied it to each row in a dataframe. 

The following is a encoding function (`encode_country`) that recieve a row as parameter, take the two most important component (country and population density, the rest of elements are not going to be used), and return the iso_code of that country and the density as a row. Noted that has a try/except block to ensure that if a country can't be read by `pycountry` don't break, and instead return just `None`.

After checked what countries were not encode because `pycountry.country.search_fuzzy` didn't recognize them, it was tried to manually do it. To do that, it was created a map or python Dict to asign these countries with their respective iso code, and then update the `encode_country` function.

In [12]:
ISO_MAP = {
    'Macau': 'MAC',
    'South Korea': 'KOR',
    'United States Virgin Islands': 'VIR',
    'North Korea': 'PRK',
    'Cape Verde': 'CPV',
    'Ivory Coast': 'CIV',
    'DR Congo': 'COD',
    'Laos': 'LAO',
    'Sint Maarten': 'SXM',
    'Curacao': 'CUW'
}


# overwrited function 
def encode_country(row):
    country, iso_code, density = (row[1], ISO_MAP[row[1]], row[2]) \
                                    if row[1] in ISO_MAP.keys() \
                                    else (row[1], None, row[2])
    try:
        iso_code = pycountry.countries.search_fuzzy(country)[0].alpha_3 \
                    if iso_code is None else None
    except LookupError: 
        pass
    return (country, iso_code, density)

In [13]:
rdd = df_pop_density.rdd.map(encode_country)
df_pop_density = rdd.toDF(['country', 'iso_code', 'density'])

# Create a new tempView for the new table with encode countries
df_pop_density.createOrReplaceTempView("population_density_encoded")

Now, the two tables can be easily compared with each other by looking at their iso_codes. Noted that there are 10 iso_code in `covid_deaths` that are not in `population_density_encode`, and that the `WHERE continent IS NOT NULL` clause is added in the SQL query beacuse when it's not passed, continents as a whole are considered as country in the dataset.

In [14]:
spark.sql("""
    SELECT iso_code 
    FROM covid_deaths
    WHERE continent IS NOT NULL
    EXCEPT
    SELECT iso_code FROM population_density_encoded
""").count()

16

Here are the countries and the iso_code from `covid_deaths` that didn't find a iso_code in the `population_density_encoded` table (the new one). There are 10 countries as it was commented before. Therefore, it can be affirmed that for these 10 countries are not population density information available in the dataset consulted.

In [15]:
spark.sql("""
    SELECT DISTINCT location, iso_code
    FROM covid_deaths
    WHERE iso_code IN (
        SELECT iso_code FROM covid_deaths
        EXCEPT
        SELECT iso_code FROM population_density_encoded) AND
        continent IS NOT NULL 
""").show()

+--------------------+--------+
|            location|iso_code|
+--------------------+--------+
|               Macao|     MAC|
|     Northern Cyprus|OWID_CYN|
|             Curacao|     CUW|
|Democratic Republ...|     COD|
|               Niger|     NER|
|         South Korea|     KOR|
|       Cote d'Ivoire|     CIV|
|            Pitcairn|     PCN|
|Sint Maarten (Dut...|     SXM|
|              Jersey|     JEY|
|          Cape Verde|     CPV|
|Bonaire Sint Eust...|     BES|
|        Saint Helena|     SHN|
|              Kosovo|OWID_KOS|
|                Laos|     LAO|
|            Guernsey|     GGY|
+--------------------+--------+



### 2. Look at the most infected countries, where are they located (which continent)

The following are the top 20 countries that have the higher rate of infected people; 
noted that neither Spain nor Italy, two of the most affected european countries at the beginning of the pandemic, appear.

In [16]:
df_top_infected = spark.sql("""
    SELECT iso_code, location, continent, population,
        MAX(total_cases) as TotalCases, 
        (MAX(total_cases)/MAX(population))*100 as PercentPopulationInfected
    FROM covid_deaths
    GROUP BY iso_code, location, population, continent
    ORDER BY PercentPopulationInfected desc
    LIMIT 20
""")
df_top_infected.createOrReplaceTempView("top_infected_countries")
df_top_infected.show()

+--------+-------------+-------------+------------+-----------+-------------------------+
|iso_code|     location|    continent|  population| TotalCases|PercentPopulationInfected|
+--------+-------------+-------------+------------+-----------+-------------------------+
|     AND|      Andorra|       Europe|     77265.0|    13900.0|       17.990034297547403|
|     MNE|   Montenegro|       Europe|    628062.0|   100230.0|       15.958615550694038|
|     SYC|   Seychelles|       Africa|     98340.0|    15579.0|        15.84197681513118|
|     BHR|      Bahrain|         Asia|   1701583.0|   265643.0|       15.611521741813359|
|     CZE|      Czechia|       Europe| 1.0708982E7|  1667115.0|       15.567446093382173|
|     SMR|   San Marino|       Europe|     33938.0|     5091.0|        15.00088396487713|
|     MDV|     Maldives|         Asia|    540542.0|    73660.0|       13.627063206929341|
|     SVN|     Slovenia|       Europe|   2078932.0|   257287.0|       12.375921867574311|
|     LUX|

The following table shows how many countries are per continent, in order to detect the most affected ones.

In [20]:
spark.sql("""
    SELECT tic.location, tic.PercentPopulationInfected, pde.density
    FROM top_infected_countries as tic
    INNER JOIN population_density_encoded as pde
        ON tic.iso_code = pde.iso_code
    ORDER BY 3 DESC
""").show()

+--------+-------------+-------------------------+---------+
|iso_code|     location|PercentPopulationInfected|  density|
+--------+-------------+-------------------------+---------+
|     AND|      Andorra|       17.990034297547403| 165.2885|
|     MNE|   Montenegro|       15.958615550694038|  45.4715|
|     SYC|   Seychelles|        15.84197681513118|  218.823|
|     BHR|      Bahrain|       15.611521741813359|2285.3542|
|     CZE|      Czechia|       15.567446093382173| 135.9862|
|     SMR|   San Marino|        15.00088396487713| 557.6557|
|     MDV|     Maldives|       13.627063206929341|1812.0567|
|     SVN|     Slovenia|       12.375921867574311| 102.5366|
|     LUX|   Luxembourg|       11.293563970503662| 245.4811|
|     SWE|       Sweden|       10.790314547487096|  22.5634|
|     URY|      Uruguay|       10.598933076778918|  19.2514|
|     SRB|       Serbia|       10.529030672798209|   98.432|
|     LTU|    Lithuania|       10.238655602946196|  41.1924|
|     USA|United States|

In [21]:
spark.sql("""
    SELECT continent, COUNT(continent), COUNT(continent)/20*100 as Percentage
    FROM top_infected_countries
    GROUP BY continent
    ORDER BY 2 desc
""").show()

+-------------+----------------+----------+
|    continent|count(continent)|Percentage|
+-------------+----------------+----------+
|       Europe|              12|      60.0|
|         Asia|               3|      15.0|
|South America|               2|      10.0|
|North America|               2|      10.0|
|       Africa|               1|       5.0|
+-------------+----------------+----------+



### 3. Look at the countries with more deaths, where are they located (which continent)

Now, let's have a look on the top 20 countries with the higher death rate

In [22]:
df_top_deaths = spark.sql("""
    SELECT location, continent, population, 
        MAX(total_cases) as TotalCases,
        MAX(total_deaths) as TotalDeaths,
        (MAX(total_deaths)/MAX(total_cases))*100 as PercentOfDeaths
    FROM covid_deaths
    GROUP BY location, population, continent
    ORDER BY PercentOfDeaths desc
    LIMIT 20
""")
df_top_deaths.createOrReplaceTempView("top_deaths_countries")
df_top_deaths.show()

+--------------------+-------------+-------------+----------+-----------+------------------+
|            location|    continent|   population|TotalCases|TotalDeaths|   PercentOfDeaths|
+--------------------+-------------+-------------+----------+-----------+------------------+
|             Vanuatu|      Oceania|     307150.0|       4.0|        1.0|              25.0|
|               Yemen|         Asia|  2.9825968E7|    6918.0|     1361.0| 19.67331598727956|
|                Peru|South America|  3.2971846E7| 2048115.0|   191899.0| 9.369542237618493|
|              Mexico|North America| 1.28932753E8| 2513164.0|   232803.0| 9.263342941407723|
|               Sudan|       Africa|  4.3849269E7|   36620.0|     2754.0| 7.520480611687602|
|               Syria|         Asia|  1.7500657E7|   25479.0|     1873.0| 7.351151929039601|
|               Egypt|       Africa| 1.02334403E8|  281031.0|    16148.0| 5.745985318345663|
|             Somalia|       Africa|  1.5893219E7|   14933.0|      775

In [23]:
spark.sql("""
    SELECT continent, COUNT(continent), COUNT(continent)/20*100 as Percentage
    FROM top_deaths_countries
    WHERE TotalDeaths/TotalCases*100 >= 3.6423
    GROUP BY continent
    ORDER BY 2 desc
""").show()

+-------------+----------------+----------+
|    continent|count(continent)|Percentage|
+-------------+----------------+----------+
|         Asia|               5|      25.0|
|       Africa|               5|      25.0|
|South America|               3|      15.0|
|       Europe|               3|      15.0|
|      Oceania|               1|       5.0|
|North America|               1|       5.0|
+-------------+----------------+----------+



### 4. Compare the previous results with Spain, Italy and USA,

Just to verify that Spain and Italy are contemplated in the dataset, let's query them. Additionally, it can be interesting to compare their infection and death rate with the US rates.

- It can be noted that US have the higher infected rate (10.12%), but the lower death rate (1.79%).
- In contrats, Italy have the lower infected rate (7.03%) and the higher death rate (2.99%).
- Finally, Spain is in the middle of the two.

In [24]:
spark.sql("""
    SELECT DISTINCT location, 
        MAX(population) as Population,
        MAX(total_cases) as TotalCases, 
        MAX(total_deaths) as TotalDeaths,
        (MAX(total_cases)/MAX(population))*100 as PercentPopulationInfected,
        (MAX(total_deaths)/MAX(total_cases))*100 as PercentOfDeaths
    FROM covid_deaths
    WHERE location LIKE 'Spain' or 
        location LIKE 'Italy' or 
        location LIKE '%tates%'
    GROUP BY location
""").show()

+-------------+------------+-----------+-----------+-------------------------+------------------+
|     location|  Population| TotalCases|TotalDeaths|PercentPopulationInfected|   PercentOfDeaths|
+-------------+------------+-----------+-----------+-------------------------+------------------+
|United States|3.31002647E8|3.3652098E7|   604474.0|        10.16671567584171| 1.796244620469131|
|        Italy| 6.0461828E7|  4259133.0|   127542.0|        7.044333823317415|2.9945531167963058|
|        Spain| 4.6754783E7|  3799733.0|    80829.0|        8.126939654494814|2.1272284131542927|
+-------------+------------+-----------+-----------+-------------------------+------------------+



Using spark.sql to query a spain dataframe

In [25]:
df_spain = spark.sql("""
    SELECT DATE(date),
        population, total_cases, new_cases, total_deaths, 
        (total_deaths/total_cases*100) as death_rate,
        (new_cases/population*100) as infected_rate
    FROM covid_deaths
    WHERE location = 'Spain' AND
        total_deaths IS NOT NULL
    ORDER by 1
""")
df_spain.createOrReplaceTempView("covid_deaths_spain")

This is how it looks like the spain dataframe, the first 20 days since is was reported the first death.

The following are some notes realted:
- The first covid death was on march (03-march-2020), there were 165 cases in total and 45 new cases.
- It's easy to see how the number of new cases and death rate are increasing.

In [26]:
df_spain.show()

+----------+-----------+-----------+---------+------------+------------------+--------------------+
|      date| population|total_cases|new_cases|total_deaths|        death_rate|       infected_rate|
+----------+-----------+-----------+---------+------------+------------------+--------------------+
|2020-03-03|4.6754783E7|      165.0|     45.0|         1.0|0.6060606060606061|9.624683746259715E-5|
|2020-03-04|4.6754783E7|      222.0|     57.0|         2.0|0.9009009009009009|1.219126607859563...|
|2020-03-05|4.6754783E7|      259.0|     37.0|         3.0|1.1583011583011582|7.913628858035766E-5|
|2020-03-06|4.6754783E7|      400.0|    141.0|         5.0|              1.25| 3.01573424049471E-4|
|2020-03-07|4.6754783E7|      500.0|    100.0|        10.0|               2.0|2.138818610279936...|
|2020-03-08|4.6754783E7|      673.0|    173.0|        17.0| 2.526002971768202| 3.70015619578429E-4|
|2020-03-09|4.6754783E7|     1073.0|    400.0|        28.0|  2.60950605778192|8.555274441119745E-4|


In [ ]:
# spark.sql("""
#     SELECT MONTH(date) AS month, YEAR(date) AS year,
#         MAX(population) as population,
#         sum(total_cases) as total_cases, 
#         sum(new_cases) as new_cases, 
#         sum(total_deaths) as total_deaths,
#         (sum(total_deaths)/sum(total_cases)*100) as death_rate,
#         (sum(total_cases)/MAX(population)*100) as infected_rate
#     FROM covid_deaths_spain
#     GROUP BY month, year
#     ORDER by year, month
# """).show()